### PR TITLE
sql: allow parsing of REASSIGN OWNED BY ... TO ...

### PR DESCRIPTION
**Pull request by angelapwen**
Part 1 of addressing #52022.

Add parsing for REASSIGN OWNED BY ... TO ... command. Plan to follow up with implementation of the command for feature parity with PostgresQL, which will resolve the issue linked.

Release note: None

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/reassign_owned [hunk 1]
```diff
@@ -0,0 +1,22 @@
+statement ok
+CREATE TABLE t()
+
+# Ensure old role must exist.
```
Comment (angelapwen) [inline pkg/sql/logictest/testdata/logic_test/reassign_owned:4]: Still need to look at the edge cases of what should/should not be possible with this command. Once I do so, can add further cases in either in this PR or the next.

File: pkg/sql/parser/sql.y [hunk 2]
```diff
@@ -7514,6 +7516,25 @@ multiple_set_clause:
     $$.val = &tree.UpdateExpr{Tuple: true, Names: $2.nameList(), Expr: $5.expr()}
   }
 
+// %Help: REASSIGN OWNED - change ownership of all objects
+// %Category: Priv
```
Comment (angelapwen) [inline pkg/sql/parser/sql.y:7520]: Is this the correct category? `GRANT` is under Priv but `ALTER` is under DDL.

File: pkg/sql/sem/tree/stmt.go [hunk 3]
```diff
@@ -568,6 +568,12 @@ func (*Prepare) StatementType() StatementType { return Ack }
 // StatementTag returns a short string identifying the type of statement.
 func (*Prepare) StatementTag() string { return "PREPARE" }
 
+// StatementType implements the Statement interface.
+func (*ReassignOwned) StatementType() StatementType { return DDL } // TODO(angelaw): Is this correct?
```
Comment (angelapwen) [inline pkg/sql/sem/tree/stmt.go:572]: Again: I believe 'Priv' is not an option here so 'DDL' is the most appropriate return type. Would appreciate any thoughts/clarifications on the statement type however.

File: pkg/sql/parser/sql.y (hunk 2 above)
Comment (otan) [reply pkg/sql/parser/sql.y:7520]: This is fine to me!

File: pkg/sql/sem/tree/stmt.go (hunk 3 above)
Comment (otan) [inline pkg/sql/sem/tree/stmt.go:572]: yep!

File: pkg/sql/parser/sql.y [hunk 4]
```diff
@@ -7514,6 +7516,25 @@ multiple_set_clause:
...
+// TODO(angelaw): Confirm category
+
+reassign_owned_stmt:
+	// TODO(angelaw): Implement and remove error
+	REASSIGN OWNED BY string_or_placeholder TO string_or_placeholder error {
+	/* SKIP DOC */
+	return unimplementedWithIssue(sqllex, 52022)
+	}
+//{
+//	$$.val = &tree.ReassignOwned{
```
Comment (otan) [inline pkg/sql/parser/sql.y:7531]: you can keep the AST node here returning the right thing (that way, you can also add parse_test.go), then return the error `unimplemented.NewWithIssue(52022, "reassign owned by us not yet implemented")` in reassigned_owned.go.

File: pkg/sql/sem/tree/stmt.go (hunk 3 above)
Comment (otan) [reply pkg/sql/sem/tree/stmt.go:572]: Yeah it's technically a DDL. I think `Priv` is just a docs thing in sql.y for the generated bnf statements.

File: pkg/sql/parser/sql.y [hunk 5]
```diff
@@ -7514,6 +7526,20 @@ multiple_set_clause:
...
+// %Help: REASSIGN OWNED - change ownership of all objects
+// %Category: Priv
+// %Text: REASSIGN OWNED BY {<name> | CURRENT_USER | SESSION_USER}[,...]
+// TO {<name> | CURRENT_USER | SESSION_USER}
+reassign_owned_stmt:
+	REASSIGN OWNED BY role_spec_list TO role_spec
+{
+	$$.val = &tree.ReassignOwned{
+      OldRoles: $4.nameList(),
+      NewRole: *newNameFromStr($6),
```
Comment (angelapwen) [inline pkg/sql/parser/sql.y:7538]: Bit of awkward syntax here.

File: pkg/sql/parser/sql.y (hunk 4 above)
Comment (angelapwen) [reply pkg/sql/parser/sql.y]: Done in newest commit!

**Comment by knz** [comment]
It's generally a costly idea to create a top level statement. There is a lot of a client code in apps that tries to be smart with the 1st keyword in SQL.

In this particular case, there's not even a good reason to need a top level statement. The logic would be better served by an ALTER sub-command. This would even make it fit the (convenient, desirable) ability combine multiple schema changes inside a single statement, which greatly helps with txn retries.

**Comment by knz** [comment]
I was just informed that pg has a top-level statement so it's all good! :100:

my point about multiple DDL inside "ALTER" remains, it is convenient.

**Comment by angelapwen** [comment]
Thank you for the review @knz! I will make it more clear that we're mimicking PostgresQL features in the commit message/PR body in the future. 

> my point about multiple DDL inside "ALTER" remains, it is convenient.
Just to make sure I understand correctly — this point is no longer relevant to this PR as it cannot be modified to an ALTER statement, right? Or is there some way to use a top-level statement here but nest its implementation in the ALTER commands?

**Comment by otan** [comment]
> Just to make sure I understand correctly — this point is no longer relevant to this PR as it cannot be modified to an ALTER statement, right? Or is there some way to use a top-level statement here but nest its implementation in the ALTER commands?

It sounds like something we should think about as a product decision. for all existing descriptors we already have `SET OWNER TO`, but if we wanted to apply everything recursively (which REASSIGN does), we'd need `ALTER DATABASE/SCHEMA x REASSIGN ...` (not table or type since they're "leaf" nodes). Maybe worth filing an issue about it (I can do so later if you feel this is accurate @knz)

**Comment by knz** [comment]
> that said, REASSIGN works on the DATABASE level so it'd be an alias for ALTER DATABASE x REASSIGN if I understand correctly .

Yes that's the way I would do it: make the ALTER form "primary implementation", then make the top-level REASSIGN an alias for it

(maybe just planning-time alias: I suspect clients care about REASSIGN keeping its statement tag in responses)

File: pkg/sql/parser/parse_test.go [hunk 6]
```diff
@@ -2595,6 +2598,9 @@ SKIP_MISSING_FOREIGN_KEYS, SKIP_MISSING_SEQUENCES, SKIP_MISSING_SEQUENCE_OWNERS,
 
 		{`ALTER TYPE t OWNER TO CURRENT_USER`, "ALTER TYPE t OWNER TO \"current_user\""},
 		{`ALTER TYPE t OWNER TO SESSION_USER`, "ALTER TYPE t OWNER TO \"session_user\""},
+
+		{`REASSIGN OWNED BY CURRENT_USER TO foo`, "REASSIGN OWNED BY \"current_user\" TO foo"},
+		{`REASSIGN OWNED BY SESSION_USER TO foo`, "REASSIGN OWNED BY \"session_user\" TO foo"},
```
Comment (otan) [inline pkg/sql/parser/parse_test.go:2603]: you can use backticks to avoid quotes, e.g.

```
`REASSIGNED OWNED BY "session_user" TO foo`
```

**Review by otan — COMMENTED** [review]
Looks good, just the failing test and we should be good!

File: pkg/sql/parser/parse_test.go (hunk 6 above)
Comment (angelapwen) [reply pkg/sql/parser/parse_test.go]: Done in most recent commit (and edited the quotes in the ALTER TYPE lines above as well) 👍

**Review by otan — APPROVED** [review]
Looks good!

File: pkg/sql/logictest/testdata/logic_test/reassign_owned (hunk 1 above)
Comment (angelapwen) [reply pkg/sql/logictest/testdata/logic_test/reassign_owned]: Will write more robust logic tests in the follow-up PR after the implementation of the command.

**Pull request merged**